tower-batch/src/worker.rs

```diff
@@ -43,66 +44,125 @@ pub(crate) struct Handle {
-        if self.finish {
-            // We've already received None and are shutting down
-            return Poll::Ready(None);
-        }
-
-        tracing::trace!("worker polling for next message");
-        if let Some(mut msg) = self.current_message.take() {
-            // poll_closed returns Poll::Ready is the receiver is dropped.
-            // Returning Pending means it is still alive, so we should still
-            // use it.
-            if msg.tx.poll_closed(cx).is_pending() {
-                tracing::trace!("resuming buffered request");
-                return Poll::Ready(Some((msg, false)));
+    async fn process_req(&mut self, req: Request, tx: message::Tx<T::Future>) {
+        if let Some(ref failed) = self.failed {
+            tracing::trace!("notifying caller about worker failure");
+            let _ = tx.send(Err(failed.clone()));
+        } else {
+            match self.service.ready_and().await {
+                Ok(svc) => {
+                    let rsp = svc.call(req.into());
+                    let _ = tx.send(Ok(rsp));
+                }
+                Err(e) => {
+                    self.failed(e.into());
+                    let _ = tx.send(Err(self
+                        .failed
+                        .as_ref()
+                        .expect("Worker::failed did not set self.failed?")
+                        .clone()));
+                }
             }
+        }
+    }
 
-            tracing::trace!("dropping cancelled buffered request");
+    async fn flush_service(&mut self) {
+        if let Err(e) = self
+            .service
+            .ready_and()
+            .and_then(|svc| svc.call(BatchControl::Flush))
+            .await
+        {
+            self.failed(e.into());
         }
+    }
 
-        // Get the next request
-        while let Some(mut msg) = ready!(Pin::new(&mut self.rx).poll_recv(cx)) {
-            if msg.tx.poll_closed(cx).is_pending() {
-                tracing::trace!("processing new request");
-                return Poll::Ready(Some((msg, true)));
+    pub async fn run(mut self) {
+        use futures::future::Either::{Left, Right};
+        // The timer is started when the first entry of a new batch is
+        // submitted, so that the batch latency of all entries is at most
+        // self.max_latency. However, we don't keep the timer running unless
+        // there is a pending request to prevent wakeups on idle services.
+        let mut timer: Option<Delay> = None;
+        let mut pending_items = 0usize;
+        loop {
+            match timer {
+                None => match self.rx.next().await {
+                    // The first message in a new batch.
+                    Some(msg) => {
+                        let span = msg.span;
+                        self.process_req(msg.request, msg.tx)
+                            // Apply the provided span to request processing
+                            .instrument(span)
```

Comment:
This ensures that the span which was current in the task where the original outer request was made is applied to the actual processing of the request here (on the worker task).